### ISSUE TITLE
Minishift fails to start on a windows box

### ISSUE THREAD

**Issue by VeerMuchandi** [minishift/minishift#888]
I am using minishift 1.0.0 

Here is what happens.

```
C:\WINDOWS\system32>minishift start --metrics
Starting local OpenShift cluster using 'hyperv' hypervisor...
Downloading ISO 'https://github.com/minishift/minishift-b2d-iso/releases/download/v1.0.2/minishift-b2d.iso'
 40.00 MB / 40.00 MB [=====================================================================================] 100.00% 0s
Downloading OpenShift binary 'oc' version 'v1.5.0'
 19.05 MB / 19.05 MB [=====================================================================================] 100.00% 0s
-- Checking OpenShift client ... OK
-- Checking Docker client ... FAIL
   Error: cannot obtain a Docker client
   Solution:

     To create a new Docker machine to run OpenShift, run this command again with
     the --create-machine argument. This will create a Docker machine named
     'openshift'.

     To use a different machine name, specify the --machine-name=NAME argument.

     If you wish to use an existing Docker machine, enable it before running this
     command by executing:

        docker-machine env

     where NAME is the name of your Docker machine.

   Caused By:
     Error: invalid endpoint
Error starting the cluster.
```

**Comment by amitkrout** [comment]
@VeerMuchandi i am unable to reproduce it on same env. Would you please let me know the docker version on your local setup ?

**Comment by VeerMuchandi** [comment]
Hi Amit,

C:\WINDOWS\system32>docker version
Client:
 Version:      17.03.1-ce
 API version:  1.27
 Go version:   go1.7.5
 Git commit:   c6d412e
 Built:        Tue Mar 28 00:40:02 2017
 OS/Arch:      windows/amd64

Server:
 Version:      17.03.1-ce
 API version:  1.27 (minimum version 1.12)
 Go version:   go1.7.5
 Git commit:   c6d412e
 Built:        Fri Mar 24 00:00:50 2017
 OS/Arch:      linux/amd64
 Experimental: true

Does docker version matter when minishift itself spins up a new vm?

Thanks
Veer

**Comment by amitkrout** [comment]
@VeerMuchandi Basically it should not be. I am asking this because i had a similar king of experience with minishift start (Windows10 + hyperv). I re installed the latest docker and then everything went fine for me. That why i am suspecting there might be a problem with docker demon already present in the local system.

**Comment by VeerMuchandi** [comment]
As suggested, I uninstalled and reinstalled Docker. Now after a long wait
after downloading ISO this is what happens

C:\WINDOWS\system32>minishift start
Starting local OpenShift cluster using 'hyperv' hypervisor...
Downloading ISO '
https://github.com/minishift/minishift-b2d-iso/releases/download/v1.0.2/minishift-b2d.iso
'
 40.00 MB / 40.00 MB
[=====================================================================================]
100.00% 0s
E0512 10:24:42.327120    6104 start.go:171] Error starting the VM: Error
creating the VM. Error creating machine: Error detecting OS: OS type not
recognized. Retrying.

**Comment by hferentschik** [comment]
@VeerMuchandi The question is what version of Windows are you using? Windows 10? It seems you also have Docker for Windows installed? Is this correct? Do you want to use Hyper-V or VirtualBox?

**Comment by VeerMuchandi** [comment]
I am using Windows 10. I am running Docker for Windows. I am using default
hyperV

**Comment by hferentschik** [comment]
If you are using Docker For Windows, your primary Hyper-V switch might be occupied by it. Did you create a switch for Minishift? See also https://docs.openshift.org/latest/minishift/getting-started/installing.html. You need to create your Hyper-V switch and then  set the environment variable HYPERV_VIRTUAL_SWITCH to the name of this variable.

**Comment by VeerMuchandi** [comment]
I created the switch as per the doc yesterday. But I am not sure what went
wrong. A couple of hour back I went back to the minishift VM and selected
that switch and then it worked. The switch has been there for two days.
Need to see if it stops working again.


On Fri, May 12, 2017 at 3:24 PM, Hardy Ferentschik <notifications@github.com

**Comment by gbraad** [comment]
This is a textbook example of an error when Docker for Windows (D4W) is installed. `oc` will check during `cluster up` for the existence of the `docker` client, as it needs this to pull the penShift Origin container. However, since the endpoint is set for the MobyLinuxVM instance, it is not acting as expected and fails. Most likely, the endpoint is set to the `dockernat` Virtual Switch, and therefore gets an IPv6 address... likely due to uninstalling, the VM has an unspecified virtual switch which caused the detection to fail (however, not sure as I haven;t seen this one before).

When Docker for Windows is installed, you ALWAYS need to specifiy the virtual switch with `HYPERV_VIRTUAL_SWITCH`, else deployment will fail because it gets an unusable IP address from the `dockernat` switch. Not sure if we need to emphasize this in the documentation. We already describe the need for the environment variable... although, maybe people are unaware that Docker for Windows creates this virtual switch. WDYT?

@VeerMuchandi If the VM can stop and start, I believe your original reported issue is resolved... as it was related to the configuration. Recently I started to notice issues with Hyper-V and the virtual networking infrastructure being unreliable (DHCP requests being unhandled and reusing old IP addresses leases on a different network). I suspect an update causes this, as it is happening for me since about a week. Doing a full Network Reset was the only way to resolve it for me. (Observed on three different machines, but as of yet unclear what causes it).

**Comment by gbraad** [comment]
@amitkrout just for reference; do we also test with a machine that has Win10+Hyper-V and Docker for Windows installed?

**Comment by amitkrout** [comment]
@gbraad yes, the windows test setup is used for manual test has the default hyper-v provider and docker installed for windows.

We have a test scenario where this command ```minishift docker-env``` need to be validated. For this scenario we need docker instance running on windows host. This combination i have tested on release candidate and nightly build.

**Comment by gbraad** [comment]
@amitkrout Great! so, this way we also cover using the `HYPERV_VIRTUAL_SWITCH` usage.

**Comment by VeerMuchandi** [comment]
@gbraad As on Fri, the issue did not come up again. So I think we can close
this.
A few things I noticed though.
- MobyLinuxVM instance starts again as soon as you start "minishift start".
- in my case, I created a virtual switch and still I had issues. All I did
was go back to minishift and clicked on the virtual switch again. Not sure
how the issue got fixed.
Will let you know if the issue repeats again.
